File: core/src/main/kotlin/br/com/dillmann/nginxignition/core/nginx/configuration/provider/AccessListFileProvider.kt
package br.com.dillmann.nginxignition.core.nginx.configuration.provider

import br.com.dillmann.nginxignition.core.accesslist.AccessList
import br.com.dillmann.nginxignition.core.accesslist.AccessListRepository
import br.com.dillmann.nginxignition.core.host.Host
import br.com.dillmann.nginxignition.core.nginx.configuration.NginxConfigurationFileProvider
import java.math.BigInteger
import java.security.MessageDigest

internal class AccessListFileProvider(
    private val accessListRepository: AccessListRepository,
): NginxConfigurationFileProvider {
    private companion object {
        private const val MD5_LENGTH = 16
        private val MD5_DIGEST = MessageDigest.getInstance("MD5")
    }

    override suspend fun provide(basePath: String, hosts: List<Host>): List<NginxConfigurationFileProvider.Output> =
        accessListRepository.findAll().flatMap { build(it, basePath) }

    private fun build(accessList: AccessList, basePath: String): List<NginxConfigurationFileProvider.Output> =
        listOfNotNull(
            buildConfFile(accessList, basePath),
            buildHtpasswdFile(accessList),
        )

    private fun buildConfFile(accessList: AccessList, basePath: String): NginxConfigurationFileProvider.Output {
        val entriesContents = accessList
            .entries
            .joinToString(separator = "\n") { entry ->
                entry
                    .sourceAddresses
                    .joinToString(separator = "\n") {
                        "${entry.outcome.toNginxOperation()} $it;"
                    }
            }

        val usernamePasswordContents =
            if (accessList.credentials.isEmpty()) ""
            else """
                auth_basic "${accessList.realm ?: "Login required"}";
                auth_basic_user_file $basePath/config/access-list-${accessList.id}.htpasswd;
            """.trimIndent()

        val satisfyContents =
            if (accessList.credentials.isEmpty() || accessList.entries.isEmpty()) "satisfy any;"
            else "satisfy ${if(accessList.satisfyAll) "all" else "any"};"

        val contents = """
            $satisfyContents
            $entriesContents
            ${accessList.defaultOutcome.toNginxOperation()} all;
            $usernamePasswordContents
        """.trimIndent()

        return NginxConfigurationFileProvider.Output(
            name = "access-list-${accessList.id}.conf",
            contents = contents,
        )
    }

    private fun buildHtpasswdFile(accessList: AccessList): NginxConfigurationFileProvider.Output? {
        if (accessList.credentials.isEmpty())
            return null

        val contents = accessList.credentials.joinToString(separator = "\n") { (username, password) ->
            val hash = MD5_DIGEST.digest(password.toByteArray())
            val stringRepresentation = BigInteger(1, hash).toString(MD5_LENGTH).padStart(MD5_LENGTH, '0')
            "$username:$stringRepresentation"
        }

        return NginxConfigurationFileProvider.Output(
            name = "access-list-${accessList.id}.htpasswd",
            contents = contents,
        )
    }

    private fun AccessList.Outcome.toNginxOperation() =
        when (this) {
            AccessList.Outcome.DENY -> "deny"
            AccessList.Outcome.ALLOW -> "allow"
        }
}

File: core/nginx/cfgfiles/access_list_file_provider.go
package cfgfiles

import (
	"fmt"
	"strings"

	"github.com/ncw/pwhash/apr1_crypt"

	"dillmann.com.br/nginx-ignition/core/accesslist"
)

type accessListFileProvider struct {
	commands accesslist.Commands
}

func newAccessListFileProvider(commands accesslist.Commands) *accessListFileProvider {
	return &accessListFileProvider{
		commands: commands,
	}
}

func (p *accessListFileProvider) provide(ctx *providerContext) ([]File, error) {
	accessLists, err := p.commands.GetAll(ctx.context)
	if err != nil {
		return nil, err
	}

	outputs := make([]File, 0)
	for _, accessList := range accessLists {
		outputs = append(outputs, p.build(&accessList, ctx.paths)...)
	}

	return outputs, nil
}

func (p *accessListFileProvider) build(accessList *accesslist.AccessList, paths *Paths) []File {
	outputs := make([]File, 0)

	if confFile := p.buildConfFile(accessList, paths); confFile != nil {
		outputs = append(outputs, *confFile)
	}

	if htpasswdFile := p.buildHtpasswdFile(accessList); htpasswdFile != nil {
		outputs = append(outputs, *htpasswdFile)
	}

	return outputs
}

func (p *accessListFileProvider) buildConfFile(
	accessList *accesslist.AccessList,
	paths *Paths,
) *File {
	entriesContents := make([]string, 0)
	for _, entry := range accessList.Entries {
		for _, sourceAddress := range entry.SourceAddress {
			entriesContents = append(
				entriesContents,
				fmt.Sprintf("%s %s;", toNginxOperation(entry.Outcome), sourceAddress),
			)
		}
	}

	usernamePasswordContents := ""
	if len(accessList.Credentials) > 0 {
		usernamePasswordContents = fmt.Sprintf(
			`
				auth_basic "%s"; 
				auth_basic_user_file %saccess-list-%s.htpasswd;
			`,
			accessList.Realm, paths.Config, accessList.ID,
		)
	}

	satisfyContents := "satisfy any;"
	if len(accessList.Credentials) > 0 && len(accessList.Entries) > 0 {
		if accessList.SatisfyAll {
			satisfyContents = "satisfy all;"
		} else {
			satisfyContents = "satisfy any;"
		}
	}

	forwardHeadersContents := ""
	if !accessList.ForwardAuthenticationHeader {
		forwardHeadersContents = `proxy_set_header Authorization "";`
	}

	contents := fmt.Sprintf(
		"%s\n%s\n%s all;\n%s\n%s",
		satisfyContents,
		strings.Join(entriesContents, "\n"),
		toNginxOperation(accessList.DefaultOutcome),
		usernamePasswordContents,
		forwardHeadersContents,
	)

	return &File{
		Name:     fmt.Sprintf("access-list-%s.conf", accessList.ID),
		Contents: contents,
	}
}

func (p *accessListFileProvider) buildHtpasswdFile(accessList *accesslist.AccessList) *File {
	if len(accessList.Credentials) == 0 {
		return nil
	}

	contents := make([]string, 0)
	for _, credential := range accessList.Credentials {
		hash := apr1_crypt.Crypt(credential.Password, apr1_crypt.GenerateSalt(8))
		contents = append(contents, fmt.Sprintf("%s:%s", credential.Username, hash))
	}

	return &File{
		Name:     fmt.Sprintf("access-list-%s.htpasswd", accessList.ID),
		Contents: strings.Join(contents, "\n"),
	}
}

func toNginxOperation(outcome accesslist.Outcome) string {
	switch outcome {
	case accesslist.AllowOutcome:
		return "allow"
	case accesslist.DenyOutcome:
		return "deny"
	default:
		return ""
	}
}
